### PR TITLE
Complete rewrite of the `fastlane init` onboarding flow

### PR DESCRIPTION
**Pull request by KrauseFx**
screenshot 2017-12-31 05 59 50

image

screenshot 2018-01-01 05 49 10

image

### CODE REVIEW

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: fastlane/lib/fastlane/setup/setup.rb [hunk 1]
```diff
@@ -1,60 +1,272 @@
...
-        UI.important("Detected react-native app. To set up fastlane, please run")
-        UI.command("fastlane init")
-        UI.important("in the sub-folder for each platform (\"ios\" or \"android\")")
-        UI.user_error!("Please navigate to the platform subfolder and run `fastlane init` again")
+        SetupAndroid.new.setup_android
+      else
+        UI.error("No iOS or Android projects found in current directory '#{Dir.pwd}'")
+        UI.error("Make sure to `cd` into a directory containing your iOS or Android app")
+        if UI.confirm("Do you still want to setup a manual fastlane config in the current directory?")
+          SetupIos.new(
```
Comment (lacostej) [inline fastlane/lib/fastlane/setup/setup.rb]: you could ask whether people want to setup ios or android by default. Right?

File: fastlane/lib/fastlane/setup/setup_android.rb [hunk 2]
```diff
@@ -1,59 +1,83 @@
...
+
+      finish_up
     end
 
-    def generate_appfile
-      UI.message('------------------------------')
-      UI.success('To not re-enter your packagename and issuer every time you run one of the fastlane tools or fastlane, these will be stored in a so-called Appfile.')
+    def fetch_information_for_appfile
+      UI.message('')
+      UI.message('To not re-enter your packagename and issuer every time you run one of fastlane, these will be stored in a so-called Appfile.')
```
Comment (lacostej) [inline fastlane/lib/fastlane/setup/setup_android.rb]: "everytime you run fastlane" ?

File: fastlane/lib/fastlane/setup/setup.rb (hunk 1 above)
Comment (KrauseFx) [reply fastlane/lib/fastlane/setup/setup.rb]: Ah nice catch, will update. Just waiting for the language checks by @taquitos 👍

File: fastlane/lib/assets/AppfileTemplate.swift [hunk 3]
```diff
@@ -1,8 +1,7 @@
 var appIdentifier: String { return "[[APP_IDENTIFIER]]" } // The bundle identifier of your app
 var appleID: String { return "[[APPLE_ID]]" } // Your Apple email address
 
-var teamID: String { return "[[DEV_PORTAL_TEAM_ID]]" } // Developer Portal Team ID
-var itcTeam: String? { return [[ITC_TEAM]] } // iTunes Connect Team ID (may be nil if no team)
+[[TEAMS]]
```
Comment (taquitos) [inline fastlane/lib/assets/AppfileTemplate.swift]: not valid swift

Comment (KrauseFx) [reply fastlane/lib/assets/AppfileTemplate.swift]: This gets replaced during onboarding, as `[[TEAMS]]` is the placeholder

Comment (KrauseFx) [reply fastlane/lib/assets/AppfileTemplate.swift]: PlaceHODLer

File: fastlane/lib/fastlane/setup/setup.rb [hunk 4]
```diff
@@ -1,60 +1,274 @@
...
+      UI.message("Generated Fastfile at path `#{fastfile_path}`")
+      UI.message("Generated Appfile at path `#{appfile_path}`")
+
+      UI.message("Please check the newly generated configuration files into git together with your project")
+      UI.message("This way, everyone in your team can easily use the fastlane setup")
+      continue_with_enter
+    end
+
+    def finish_up
+      write_fastfile!
```
Comment (taquitos) [inline fastlane/lib/fastlane/setup/setup.rb]: so declarative, best 🎈

File: fastlane/lib/fastlane/setup/setup.rb [hunk 5]
```diff
@@ -1,60 +1,274 @@
...
         return
       end
 
-      platform = nil
-      if is_ios?
-        UI.message("Detected iOS/Mac project in current directory...")
-        platform = :ios
-      elsif is_android?
+      # this is used by e.g. configuration.rb to not show warnings when running produce
+      ENV["FASTLANE_ONBOARDING_IN_PROCESS"] = 1.to_s
```
Comment (taquitos) [inline fastlane/lib/fastlane/setup/setup.rb]: global variables are the best

**Review by hjanuschka — APPROVED** [review]

**Pull request merged**